File: src/errors/PaymentAuthorizationConfigurationError.js
'use strict';

/**
 * Raised when a card-network adapter is handed an authorisation profile it is not
 * certified to use. This is a configuration fault, not a customer or issuer outcome,
 * so it surfaces as HTTP 500 rather than a decline.
 */
class PaymentAuthorizationConfigurationError extends Error {
  constructor(message, context = {}) {
    super(message);
    this.name = 'PaymentAuthorizationConfigurationError';
    this.cardNetwork = context.cardNetwork || null;
    this.selectedProfile = context.selectedProfile || null;
    this.expectedProfile = context.expectedProfile || null;
    this.httpStatus = 500;
    Error.captureStackTrace(this, PaymentAuthorizationConfigurationError);
  }

  toLogFields() {
    return {
      error: this.name,
      card_network: this.cardNetwork,
      profile: this.selectedProfile,
      expected_profile: this.expectedProfile,
    };
  }
}

module.exports = { PaymentAuthorizationConfigurationError };

File: src/main/kotlin/com/beaconstone/payments/errors/PaymentAuthorizationConfigurationError.kt
package com.beaconstone.payments.errors

/**
 * Raised when a card-network adapter is handed an authorisation profile it is not
 * certified to use. This is a configuration fault, not a customer or issuer outcome,
 * so it surfaces as HTTP 500 rather than a decline.
 */
class PaymentAuthorizationConfigurationError(
    message: String,
    val cardNetwork: String? = null,
    val selectedProfile: String? = null,
    val expectedProfile: String? = null,
) : RuntimeException(message) {
    val httpStatus: Int = 500
    val errorName: String = "PaymentAuthorizationConfigurationError"

    fun toResponse(): Map<String, Any?> = mapOf(
        "error" to errorName,
        "message" to message,
        "cardNetwork" to cardNetwork,
        "selectedProfile" to selectedProfile,
        "expectedProfile" to expectedProfile,
    )
}
